harness/determined/searcher/search_method.py

```diff
@@ -0,0 +1,245 @@
+import dataclasses
+import json
+import pathlib
+import uuid
+from abc import abstractmethod
+from enum import Enum
+from typing import Any, Dict, List, Optional, Set
+
+from determined.common.api import bindings
+from determined.common.experimental import Checkpoint
+
+STATE_FILE = "state"
+
+
+@dataclasses.dataclass
+class SearcherState:
+    failures: Set[uuid.UUID]
+    trial_progress: Dict[uuid.UUID, float]
+    trials_closed: Set[uuid.UUID]
+    trials_created: Set[uuid.UUID]
+    last_event_id: int = 0
+    experiment_completed: bool = False
+
+    def __init__(self) -> None:
+        self.failures = set()
+        self.trial_progress = {}
+        self.trials_closed = set()
+        self.trials_created = set()
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "failures": [str(f) for f in self.failures],
+            "trialProgress": {str(k): v for k, v in self.trial_progress.items()},
+            "trialsClosed": [str(t) for t in self.trials_closed],
+            "trialsCreated": [str(t) for t in self.trials_created],
+            "lastEventId": self.last_event_id,
+            "experimentId": self.experiment_id,
+            "experimentCompleted": self.experiment_completed,
+        }
+
+    def from_dict(self, d: Dict[str, Any]) -> None:
+        self.failures = {uuid.UUID(f) for f in d.get("failures", [])}
+        self.trial_progress = {uuid.UUID(k): v for k, v in d.get("trialProgress", {}).items()}
+        self.trials_closed = {uuid.UUID(t) for t in d.get("trialsClosed", [])}
+        self.trials_created = {uuid.UUID(t) for t in d.get("trialsCreated", [])}
+        self.last_event_id = d.get("lastEventId", 0)
+        self.experiment_id = d.get("experimentId")
+        self.experiment_completed = d.get("experimentCompleted", False)
+
+
+class ExitedReason(Enum):
+    ERRORED = "ERRORED"
+    USER_CANCELED = "USER_CANCELED"
+    INVALID_HP = "INVALID_HP"
+
+    @classmethod
+    def _from_bindings(
+        cls, bindings_exited_reason: bindings.v1TrialExitedEarlyExitedReason
+    ) -> "ExitedReason":
+        if (
+            bindings_exited_reason
+            == bindings.v1TrialExitedEarlyExitedReason.EXITED_REASON_INVALID_HP
+        ):
+            return cls.INVALID_HP
+        if (
+            bindings_exited_reason
+            == bindings.v1TrialExitedEarlyExitedReason.EXITED_REASON_USER_REQUESTED_STOP
+        ):
+            return cls.USER_CANCELED
+        raise RuntimeError(f"Invalid exited reason: {bindings_exited_reason}")
```

Comment:
Can this get rewritten to look more like the [python sdk enums](https://github.com/determined-ai/determined/blob/master/harness/determined/common/experimental/experiment.py#L12)?

Reply:
The `bindings.v1TrialExitedEarlyExitedReason` are internal whereas `ExitedReason` values are user-facing and more user-friendly. I realize we missed `EXITED_REASON_UNSPECIFIED`, and I am going to add it. However, I think it is better to have short descriptive values like "ERRORED" as opposed to "EXITED_REASON_UNSPECIFIED".

Reply:
EXITED_REASON_UNSPECIFIED probably shouldn't be part of it... that's sort of the "not present" value in protobuf, and we should probably just raise an error rather than hand the user something invalid.